File: 54.spiral-matrix.js
/*
 * @lc app=leetcode id=54 lang=javascript
 *
 * [54] Spiral Matrix
 */

// @lc code=start
/**
 * @param {number[][]} matrix
 * @return {number[]}
 */
var spiralOrder = function (matrix) {
  const res = [];
  let [left, right] = [0, matrix[0].length - 1];
  let [top, bottom] = [0, matrix.length - 1];
  while (left <= right && top <= bottom) {
    for (let i = left; i < right + 1; ++i) {
      res.push(matrix[top][i]);
    }
    if (++top > bottom) {
      break;
    }

    for (let i = top; i < bottom + 1; ++i) {
      res.push(matrix[i][right]);
    }
    if (--right < left) {
      break;
    }

    for (let i = right; i > left - 1; --i) {
      res.push(matrix[bottom][i]);
    }
    if (--bottom < top) {
      break;
    }

    for (let i = bottom; i > top - 1; --i) {
      res.push(matrix[i][left]);
    }
    if (++left > right) {
      break;
    }
  }
  return res;
};
// @lc code=end

File: 54.spiral-matrix.py
#
# @lc app=leetcode id=54 lang=python3
#
# [54] Spiral Matrix
#

# Time:  O(m * n)
# Space: O(1)

from typing import List


# @lc code=start
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        if not matrix or not matrix[0]:
            return []

        res, left, right, top, bottom = [], 0, len(matrix[0]) - 1, 0, len(matrix) - 1
        while left <= right and top <= bottom:
            for j in range(left, right + 1):
                res.append(matrix[top][j])
            top += 1
            if top > bottom:
                break

            for i in range(top, bottom + 1):
                res.append(matrix[i][right])
            right -= 1
            if right < left:
                break

            for j in range(right, left - 1, -1):
                res.append(matrix[bottom][j])
            bottom -= 1
            if bottom < top:
                break

            for i in range(bottom, top - 1, -1):
                res.append(matrix[i][left])
            left += 1
            if left > right:
                break

        return res


# @lc code=end
